File: node/createUser.js
const { v4: uuidv4 } = require('uuid');
const db = require('./utils/dbClient');

exports.handler = async (event) => {
  const { name, email } = JSON.parse(event.body);
  
  const emailCheck = await db.query({
    TableName: 'Users',
    IndexName: 'EmailIndex',
    KeyConditionExpression: 'email = :email',
    ExpressionAttributeValues: {
      ':email': email
    }
  }).promise();

  if(emailCheck.Items.length > 0) {
    return {
      statusCode: 400,
      body: JSON.stringify({ message: 'Email already in use' })
    };
  }

  // 2. Create the new user
  const user = {
    userId: uuidv4(),
    name,
    email,
    createdAt: new Date().toISOString()
  };

  await db.put({ TableName: 'Users', Item: user }).promise();

  return {
    statusCode: 201,
    body: JSON.stringify(user),
  };
};

File: python/create_user.py
import json
import uuid
import boto3
from datetime import datetime
from boto3.dynamodb.conditions import Key

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table('Users')

def lambda_handler(event, context):
    data = json.loads(event['body'])
    name = data.get('name')
    email = data.get('email')

    if not name or not email:
        return {
            "statusCode": 400,
            "body": json.dumps({"message": "Name and email are required"})
        }

    # Check email uniqueness using GSI
    response = table.query(
        IndexName='EmailIndex',
        KeyConditionExpression=Key('email').eq(email)
    )

    if response['Items']:
        return {
            "statusCode": 400,
            "body": json.dumps({"message": "Email already in use"})
        }

    user = {
        'userId': str(uuid.uuid4()),
        'name': name,
        'email': email,
        'createdAt': datetime.utcnow().isoformat()
    }

    table.put_item(Item=user)

    return {
        "statusCode": 201,
        "body": json.dumps(user)
    }
